[ERROR] Error raised: 0x00

Hi,
I'm done implementing the LED functionality, but the problem is that now I get an error every time I run my example `[ERROR] Error raised: 0x00` although everything works just fine. Could you help me out? I also don't know how to make a pull request, never done that before 😅 Do you have a discord server that I can join? Or can you explain a bit how this works?


PS: Your code is really well organized and well commented. I was able to understand the structure in less than 10 minutes (although I have limited experience in C). I'm impressed :)

--- a/examples/example_led.c
+++ b/examples/example_led.c
@@ -51,7 +51,6 @@ int main(int argc, char *argv[]) {
         return -1;
     }
 
-    errno = 0;
     speed = strtol(argv[3], &endptr, 10);
 
     if (errno != 0 || *endptr != '\0' || speed > 255 || speed < 0) {
@@ -59,7 +58,6 @@ int main(int argc, char *argv[]) {
         return -1;
     }
 
-    errno = 0;
     color = strtol(argv[4], &endptr, 10);
 
     if (errno != 0 || *endptr != '\0' || color < 1 || color > 3) {
@@ -67,7 +65,6 @@ int main(int argc, char *argv[]) {
         return -1;
     }
 
-    errno = 0;
     times = strtol(argv[5], &endptr, 10);
 
     if (errno != 0 || *endptr != '\0' || times > 255 || times < 0) {

--- a/examples/example_led.c
+++ b/examples/example_led.c
@@ -2,9 +2,7 @@
 #include <stdio.h>
 #include <string.h>
 #include <stdlib.h>
-#include <limits.h>
 #include <errno.h>
-
 
 int main(int argc, char *argv[]) {
     Driver *driver;
@@ -12,19 +10,23 @@ int main(int argc, char *argv[]) {
     int32_t err = SUCCESS;
 
     if (argc == 2 && strncmp(argv[1], "--help", 6) == 0) {
-        printf("Usage: %s <port> <control> <speed> <color> <times>\n", argv[0]);
-        printf("port - your port, ex: \"/dev/ttyS0\" if you're using UART on RaspberryPi\n");
-        printf("control   1: breathing light\n");
-        printf("          2: flashing light\n");
-        printf("          3: light always on\n");
-        printf("          4: light always off\n");
-        printf("          5: light gradually on\n");
-        printf("          6: light gradually off\n");
-        printf("speed     0-255 It is effective for breathing lamp and flashing lamp,Light gradually on,Light gradually off\n");
-        printf("color     1: red\n");
-        printf("          2: blue\n");
-        printf("          3: purple\n");
-        printf("times     1-255, 0 for infinite\n");
+        printf("Usage: %s <port> <control> <speed> <color> <times>\n"
+        "port - UART communication device\n"
+        "control:\n"
+        "     1 - breathing light\n"
+        "     2 - flashing light\n"
+        "     3 - light always on\n"
+        "     4 - light always off\n"
+        "     5 - light gradually on\n"
+        "     6 - light gradually off\n"
+        "speed:\n"
+        "     0-255 - It is effective for breathing lamp and flashing lamp,Light gradually on,Light gradually off\n"
+        "color:\n"
+        "     1 - red\n"
+        "     2 - blue\n"
+        "     3 - purple\n"
+        "times:\n"
+        "     1-255, 0 for infinite\n", argv[0]);
         return 0;
     }
 
@@ -33,79 +35,43 @@ int main(int argc, char *argv[]) {
         return -1;
     }
 
-    char *controlPointer;
-    char *speedPointer;
-    char *colorPointer;
-    char *timesPointer;
+    int8_t *endptr;
 
-    int control;
-    int speed;
-    int color;
-    int times;
+    int32_t control;
+    int32_t speed;
+    int32_t color;
+    int32_t times;
 
     errno = 0;
-    long controlConv = strtol(argv[2], &controlPointer, 10);
+    control = strtol(argv[2], &endptr, 10);
 
     // Check for errors: e.g., the string does not represent an integer
-    // or the integer is larger than int
-    if (errno != 0 || *controlPointer != '\0' || controlConv > INT_MAX || controlConv < INT_MIN) {
-        printf("[ERROR] The option for the control is not an integer!\n");
+    if (errno != 0 || *endptr != '\0' || control < 1 || control > 6) {
+        printf("[ERROR] The option for the control is not valid, see %s --help for help\n", argv[0]);
         return -1;
-    } else {
-        control = controlConv;
     }
 
     errno = 0;
-    long speedConv   = strtol(argv[3], &speedPointer, 10);
+    speed = strtol(argv[3], &endptr, 10);
 
-    if (errno != 0 || *speedPointer != '\0' || speedConv > INT_MAX || speedConv < INT_MIN) {
-        printf("[ERROR] The option for the speed is not an integer!\n");
+    if (errno != 0 || *endptr != '\0' || speed > 255 || speed < 0) {
+        printf("[ERROR] The option for the speed is not valid, see %s --help for help\n", argv[0]);
         return -1;
-    } else {
-        speed = speedConv;
     }
 
     errno = 0;
-    long colorConv   = strtol(argv[4], &colorPointer, 10);
+    color = strtol(argv[4], &endptr, 10);
 
-    if (errno != 0 || *colorPointer != '\0' || colorConv > INT_MAX || colorConv < INT_MIN) {
-        printf("[ERROR] The option for the color is not an integer!\n");
+    if (errno != 0 || *endptr != '\0' || color < 1 || color > 3) {
+        printf("[ERROR] The option for the color is not valid, see %s --help for help\n", argv[0]);
         return -1;
-    } else {
-        color = colorConv;
     }
 
     errno = 0;
-    long timesConv   = strtol(argv[5], &timesPointer, 10);
+    times = strtol(argv[5], &endptr, 10);
 
-    if (errno != 0 || *timesPointer != '\0' || timesConv > INT_MAX || timesConv < INT_MIN) {
-        printf("[ERROR] The option for times is not an integer!\n");
-        return -1;
-    } else {
-        times = timesConv;
-    }
-
-    // Checking control variable
-    if (control != 1 && control != 2 && control != 3 && control != 4 && control != 5 && control != 6) {
-        printf("Control variable not set properly see %s --help for help\n", argv[0]);
-        return -1;
-    }
-
-    // Checking speed variable
-    if (speed > 255 || speed < 0) {
-        printf("Speed variable not set properly see %s --help for help\n", argv[0]);
-        return -1;
-    }
-    
-    //Checking color variable
-    if (color != 1 && color != 2 && color != 3) {
-        printf("Color variable not set properly see %s --help for help\n", argv[0]);
-        return -1;
-    }
-
-    // Checking times variable
-    if (times > 255 || times < 0) {
-        printf("Times variable not set properly see %s --help for help\n", argv[0]);
+    if (errno != 0 || *endptr != '\0' || times > 255 || times < 0) {
+        printf("[ERROR] The option for times is not valid, see %s --help for help\n", argv[0]);
         return -1;
     }
 
@@ -125,10 +91,11 @@ int main(int argc, char *argv[]) {
     if (err != SUCCESS)
         goto error;
 
-    printf("[INFO]  Executing command with the following parameters: %d %d %d %d\n", control, speed, color, times);
+    printf("[INFO]  Executing command: %s %d %d %d %d\n", argv[0], control, speed, color, times);
     err = call_cmd(driver, AuraLedConfig, &reply, 4, control, speed, color, times);
     if (err != SUCCESS)
         goto error;
+    return 0;
 
 error:
     printf("[ERROR] Error raised: 0x%.2X\n", err);

--- a/examples/example_led.c
+++ b/examples/example_led.c
@@ -1,0 +1,136 @@
+#include <r502.h>
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <limits.h>
+#include <errno.h>
+
+
+int main(int argc, char *argv[]) {
+    Driver *driver;
+    Reply reply;
+    int32_t err = SUCCESS;
+
+    if (argc == 2 && strncmp(argv[1], "--help", 6) == 0) {
+        printf("Usage: %s <port> <control> <speed> <color> <times>\n", argv[0]);
+        printf("port - your port, ex: \"/dev/ttyS0\" if you're using UART on RaspberryPi\n");
+        printf("control   1: breathing light\n");
+        printf("          2: flashing light\n");
+        printf("          3: light always on\n");
+        printf("          4: light always off\n");
+        printf("          5: light gradually on\n");
+        printf("          6: light gradually off\n");
+        printf("speed     0-255 It is effective for breathing lamp and flashing lamp,Light gradually on,Light gradually off\n");
+        printf("color     1: red\n");
+        printf("          2: blue\n");
+        printf("          3: purple\n");
+        printf("times     1-255, 0 for infinite\n");
+        return 0;
+    }
+
+    if (argc != 6) {
+        printf("Usage: %s <port> <control> <speed> <color> <times>\nUse %s --help for help\n", argv[0], argv[0]);
+        return -1;
+    }
+
+    char *controlPointer;
+    char *speedPointer;
+    char *colorPointer;
+    char *timesPointer;
+
+    int control;
+    int speed;
+    int color;
+    int times;
+
+    errno = 0;
+    long controlConv = strtol(argv[2], &controlPointer, 10);
+
+    // Check for errors: e.g., the string does not represent an integer
+    // or the integer is larger than int
+    if (errno != 0 || *controlPointer != '\0' || controlConv > INT_MAX || controlConv < INT_MIN) {
+        printf("[ERROR] The option for the control is not an integer!\n");
+        return -1;
+    } else {
+        control = controlConv;
+    }
+
+    errno = 0;
+    long speedConv   = strtol(argv[3], &speedPointer, 10);
+
+    if (errno != 0 || *speedPointer != '\0' || speedConv > INT_MAX || speedConv < INT_MIN) {
+        printf("[ERROR] The option for the speed is not an integer!\n");
+        return -1;
+    } else {
+        speed = speedConv;
+    }
+
+    errno = 0;
+    long colorConv   = strtol(argv[4], &colorPointer, 10);
+
+    if (errno != 0 || *colorPointer != '\0' || colorConv > INT_MAX || colorConv < INT_MIN) {
+        printf("[ERROR] The option for the color is not an integer!\n");
+        return -1;
+    } else {
+        color = colorConv;
+    }
+
+    errno = 0;
+    long timesConv   = strtol(argv[5], &timesPointer, 10);
+
+    if (errno != 0 || *timesPointer != '\0' || timesConv > INT_MAX || timesConv < INT_MIN) {
+        printf("[ERROR] The option for times is not an integer!\n");
+        return -1;
+    } else {
+        times = timesConv;
+    }
+
+    // Checking control variable
+    if (control != 1 && control != 2 && control != 3 && control != 4 && control != 5 && control != 6) {
+        printf("Control variable not set properly see %s --help for help\n", argv[0]);
+        return -1;
+    }
+
+    // Checking speed variable
+    if (speed > 255 || speed < 0) {
+        printf("Speed variable not set properly see %s --help for help\n", argv[0]);
+        return -1;
+    }
+    
+    //Checking color variable
+    if (color != 1 && color != 2 && color != 3) {
+        printf("Color variable not set properly see %s --help for help\n", argv[0]);
+        return -1;
+    }
+
+    // Checking times variable
+    if (times > 255 || times < 0) {
+        printf("Times variable not set properly see %s --help for help\n", argv[0]);
+        return -1;
+    }
+
+    printf("[INFO]  Using port %s\n", argv[1]);
+    printf("[INFO]  Initializing device communication\n");
+    err = init_driver(argv[1], 0xFFFFFFFF, &driver);
+    if (err != SUCCESS)
+        goto error;
+
+    printf("[INFO]  Checking device status\n");
+    err = call_cmd(driver, ReadSysPara, &reply, 0);
+    if (err != SUCCESS)
+        goto error;
+
+    printf("[INFO]  Verifying password (default: 0x00000000)\n");
+    err = call_cmd(driver, VfyPwd, &reply, 1, 0x00000000);
+    if (err != SUCCESS)
+        goto error;
+
+    printf("[INFO]  Executing command with the following parameters: %d %d %d %d\n", control, speed, color, times);
+    err = call_cmd(driver, AuraLedConfig, &reply, 4, control, speed, color, times);
+    if (err != SUCCESS)
+        goto error;
+
+error:
+    printf("[ERROR] Error raised: 0x%.2X\n", err);
+    return err;
+}

--- a/include/r502_commands.h
+++ b/include/r502_commands.h
@@ -8,20 +8,21 @@
 #define CHECKSUM_LEN  2
 
 /* Command package lengths */
-#define READ_SYS_PARA_LEN 12
-#define VFY_PWD_LEN       16
-#define GEN_IMG_LEN       12
-#define IMG2TZ_LEN        13
-#define SEARCH_LEN        17
-#define LOAD_CHAR_LEN     15
-#define MATCH_LEN         12
-#define TEMPLATE_NUM_LEN  12
-#define REG_MODEL_LEN     12
-#define STORE_LEN         15
-#define DELETE_CHAR_LEN   16
-#define UP_CHAR_LEN       13
-#define WRITE_NOTEPAD_LEN 45
-#define READ_NOTEPAD_LEN  13
+#define READ_SYS_PARA_LEN   12
+#define VFY_PWD_LEN         16
+#define GEN_IMG_LEN         12
+#define IMG2TZ_LEN          13
+#define SEARCH_LEN          17
+#define LOAD_CHAR_LEN       15
+#define MATCH_LEN           12
+#define TEMPLATE_NUM_LEN    12
+#define REG_MODEL_LEN       12
+#define STORE_LEN           15
+#define DELETE_CHAR_LEN     16
+#define UP_CHAR_LEN         13
+#define WRITE_NOTEPAD_LEN   45
+#define READ_NOTEPAD_LEN    13
+#define AURA_LED_CONFIG_LEN 16 
 
 
 /* Build a package to send for command using the given arguments */
@@ -62,6 +63,9 @@ static int32_t reg_model_pkg(uint8_t *pkg, Command command, int32_t pkg_len);
 
 /* Build package for Store command */
 static int32_t store_pkg(uint8_t *pkg, Command command, int32_t pkg_len);
+
+/* Build package for AuraLedConfig */
+static int32_t aura_led_config(uint8_t *pkg, Command command, int32_t pkg_len);
 
 /* Build package for DeleteChar command */
 static int32_t delete_char_pkg(uint8_t *pkg, Command command, int32_t pkg_len);

--- a/include/r502_commands.h
+++ b/include/r502_commands.h
@@ -64,9 +64,6 @@ static int32_t reg_model_pkg(uint8_t *pkg, Command command, int32_t pkg_len);
 /* Build package for Store command */
 static int32_t store_pkg(uint8_t *pkg, Command command, int32_t pkg_len);
 
-/* Build package for AuraLedConfig */
-static int32_t aura_led_config(uint8_t *pkg, Command command, int32_t pkg_len);
-
 /* Build package for DeleteChar command */
 static int32_t delete_char_pkg(uint8_t *pkg, Command command, int32_t pkg_len);
 
@@ -79,5 +76,7 @@ static int32_t write_notepad_pkg(uint8_t *pkg, Command command, int32_t pkg_len)
 /* Build package for ReadNotepad command */
 static int32_t read_notepad_pkg(uint8_t *pkg, Command command, int32_t pkg_len);
 
+/* Build package for AuraLedConfig */
+static int32_t aura_led_config(uint8_t *pkg, Command command, int32_t pkg_len);
 
 #endif

--- a/include/r502_reply.h
+++ b/include/r502_reply.h
@@ -7,20 +7,21 @@
 #define MAX_DATA_LENGTH 128
 
 /* Reply package lengths */
-#define READ_SYS_PARA_REPLY_LEN  28
-#define VFY_PWD_REPLY_LEN        12
-#define GEN_IMG_REPLY_LEN        12
-#define IMG2TZ_REPLY_LEN         12
-#define SEARCH_REPLY_LEN         16
-#define LOAD_CHAR_REPLY_LEN      12
-#define MATCH_REPLY_LEN          14
-#define TEMPLATE_NUM_REPLY_LEN   14
-#define REG_MODEL_REPLY_LEN      12
-#define STORE_REPLY_LEN          12
-#define DELETE_CHAR_REPLY_LEN    12
-#define UP_CHAR_REPLY_LEN        12
-#define WRITE_NOTEPAD_REPLY_LEN  12
-#define READ_NOTEPAD_REPLY_LEN   44
+#define READ_SYS_PARA_REPLY_LEN   28
+#define VFY_PWD_REPLY_LEN         12
+#define GEN_IMG_REPLY_LEN         12
+#define IMG2TZ_REPLY_LEN          12
+#define SEARCH_REPLY_LEN          16
+#define LOAD_CHAR_REPLY_LEN       12
+#define MATCH_REPLY_LEN           14
+#define TEMPLATE_NUM_REPLY_LEN    14
+#define REG_MODEL_REPLY_LEN       12
+#define STORE_REPLY_LEN           12
+#define AURA_LED_CONFIG_REPLY_LEN 12
+#define DELETE_CHAR_REPLY_LEN     12
+#define UP_CHAR_REPLY_LEN         12
+#define WRITE_NOTEPAD_REPLY_LEN   12
+#define READ_NOTEPAD_REPLY_LEN    44
 
 /* Additional reply package lengths */
 #define UP_CHAR_ADDITIONAL_REPLY_LEN  1536

--- a/include/r502_types.h
+++ b/include/r502_types.h
@@ -69,7 +69,9 @@ typedef enum CommandType {
 
     WriteNotepad,
 
-    ReadNotepad
+    ReadNotepad,
+
+    AuraLedConfig
 
 } CommandType;
 
@@ -153,6 +155,35 @@ typedef struct Command {
         struct ReadNotepad {
             uint8_t page_num;
         } read_notepad;
+
+        struct AuraLedConfig {
+            /*
+                Control code:
+                0x01 - breathing light
+                0x02 - flashing light
+                0x03 - light always on
+                0x04 - light always off
+                0x05 - light gradually on
+                0x06 - light gradually off
+            */
+            uint8_t ctrl;
+            /* Speed: 0x00-0xff, 256 gears, minimum 5s cycle */
+            uint8_t speed;
+            /*
+                Colors:
+                0x01 - Red
+                0x02 - Blue
+                0x03 - Purple
+            */
+            uint8_t color;
+            /* 
+                Number of cycles:
+                0 - infinite
+                1-255
+            */
+            uint8_t times;
+        } aura_led_config;
+        
 
     } body;
 
@@ -247,6 +278,9 @@ typedef struct Reply {
         struct ReadDataReply {
             uint8_t data[PAGE_SIZE];
         } read_notepad;
+
+        struct AuraLedConfigReply {} aura_led_config;
+        
 
     } body;
 

--- a/include/r502_types.h
+++ b/include/r502_types.h
@@ -71,6 +71,7 @@ typedef enum CommandType {
 
     ReadNotepad,
 
+    /* Control the LED light of the sensor */
     AuraLedConfig
 
 } CommandType;
@@ -167,8 +168,10 @@ typedef struct Command {
                 0x06 - light gradually off
             */
             uint8_t ctrl;
+            
             /* Speed: 0x00-0xff, 256 gears, minimum 5s cycle */
             uint8_t speed;
+            
             /*
                 Colors:
                 0x01 - Red
@@ -176,6 +179,7 @@ typedef struct Command {
                 0x03 - Purple
             */
             uint8_t color;
+            
             /* 
                 Number of cycles:
                 0 - infinite

--- a/src/r502_commands.c
+++ b/src/r502_commands.c
@@ -6,7 +6,7 @@
 static int32_t (*cmd_func[])(uint8_t *, Command, int32_t) = {
     read_sys_para_pkg, vfy_pwd_pkg, gen_img_pkg, img2tz_pkg, search_pkg,
     load_char_pkg, match_pkg, template_num_pkg, reg_model_pkg, store_pkg,
-    delete_char_pkg, up_char_pkg, write_notepad_pkg, read_notepad_pkg
+    delete_char_pkg, up_char_pkg, write_notepad_pkg, read_notepad_pkg, aura_led_config
 };
 
 int32_t get_basic_header(Driver *driver, uint8_t **basic_header) {
@@ -139,6 +139,15 @@ int32_t populate_command_args(Command *command, int32_t arg_num, va_list ap) {
 
         command->body.store.buf = va_arg(ap, uint32_t);
         command->body.store.index = va_arg(ap, uint32_t);
+
+        break;
+    case AuraLedConfig:
+        if (arg_num < 4)
+            goto error;
+        command->body.aura_led_config.ctrl = va_arg(ap, uint32_t);
+        command->body.aura_led_config.speed = va_arg(ap, uint32_t);
+        command->body.aura_led_config.color = va_arg(ap, uint32_t);
+        command->body.aura_led_config.times = va_arg(ap, uint32_t);
 
         break;
     case DeleteChar:
@@ -470,6 +479,48 @@ static int32_t store_pkg(uint8_t *pkg, Command command, int32_t pkg_len) {
 
     return SUCCESS;
 }
+
+
+static int32_t aura_led_config(uint8_t *pkg, Command command, int32_t pkg_len) {
+    // Required packet:
+    // basic_header                         [7]
+    // length    | 0x00 0x07                [2]
+    // instr     | 0x35                     [1]
+    // control   | aura_led_config.ctrl     [1]
+    // speed     | aura_led_config.speed    [1]
+    // color     | aura_led_config.color    [1]
+    // times     | aura_led_config.times    [1]
+    // chksum    | checksum                 [2]
+    /* Package length */
+    pkg[7] = 0x00;
+    pkg[8] = 0x07;
+
+    /* Instruction code */
+    pkg[9] = 0x35;
+
+    /* Control code */
+    pkg[10] = command.body.aura_led_config.ctrl;
+
+    /* Speed */
+    pkg[11] = command.body.aura_led_config.speed;
+
+    /* Color */
+    pkg[12] = command.body.aura_led_config.color;
+
+    /* Times */
+    pkg[13] = command.body.aura_led_config.times;
+
+    /* Checksum */
+    uint16_t chk = checksum(pkg, 6, pkg_len - 2);
+    uint8_t *chk_bytes = to_bytes_MSB(&chk, CHECKSUM_LEN);
+    pkg[14] = chk_bytes[0];
+    pkg[15] = chk_bytes[1];
+
+    free(chk_bytes);
+
+    return SUCCESS;
+}
+
 
 static int32_t delete_char_pkg(uint8_t *pkg, Command command, int32_t pkg_len) {
     // Required packet:

--- a/src/r502_commands.c
+++ b/src/r502_commands.c
@@ -141,15 +141,6 @@ int32_t populate_command_args(Command *command, int32_t arg_num, va_list ap) {
         command->body.store.index = va_arg(ap, uint32_t);
 
         break;
-    case AuraLedConfig:
-        if (arg_num < 4)
-            goto error;
-        command->body.aura_led_config.ctrl = va_arg(ap, uint32_t);
-        command->body.aura_led_config.speed = va_arg(ap, uint32_t);
-        command->body.aura_led_config.color = va_arg(ap, uint32_t);
-        command->body.aura_led_config.times = va_arg(ap, uint32_t);
-
-        break;
     case DeleteChar:
         if (arg_num < 2)
             goto error;
@@ -183,6 +174,16 @@ int32_t populate_command_args(Command *command, int32_t arg_num, va_list ap) {
         break;
     default:
         goto error;
+
+    case AuraLedConfig:
+        if (arg_num < 4)
+            goto error;
+        command->body.aura_led_config.ctrl = va_arg(ap, uint32_t);
+        command->body.aura_led_config.speed = va_arg(ap, uint32_t);
+        command->body.aura_led_config.color = va_arg(ap, uint32_t);
+        command->body.aura_led_config.times = va_arg(ap, uint32_t);
+
+        break;
     }
 
     return SUCCESS;
@@ -480,48 +481,6 @@ static int32_t store_pkg(uint8_t *pkg, Command command, int32_t pkg_len) {
     return SUCCESS;
 }
 
-
-static int32_t aura_led_config(uint8_t *pkg, Command command, int32_t pkg_len) {
-    // Required packet:
-    // basic_header                         [7]
-    // length    | 0x00 0x07                [2]
-    // instr     | 0x35                     [1]
-    // control   | aura_led_config.ctrl     [1]
-    // speed     | aura_led_config.speed    [1]
-    // color     | aura_led_config.color    [1]
-    // times     | aura_led_config.times    [1]
-    // chksum    | checksum                 [2]
-    /* Package length */
-    pkg[7] = 0x00;
-    pkg[8] = 0x07;
-
-    /* Instruction code */
-    pkg[9] = 0x35;
-
-    /* Control code */
-    pkg[10] = command.body.aura_led_config.ctrl;
-
-    /* Speed */
-    pkg[11] = command.body.aura_led_config.speed;
-
-    /* Color */
-    pkg[12] = command.body.aura_led_config.color;
-
-    /* Times */
-    pkg[13] = command.body.aura_led_config.times;
-
-    /* Checksum */
-    uint16_t chk = checksum(pkg, 6, pkg_len - 2);
-    uint8_t *chk_bytes = to_bytes_MSB(&chk, CHECKSUM_LEN);
-    pkg[14] = chk_bytes[0];
-    pkg[15] = chk_bytes[1];
-
-    free(chk_bytes);
-
-    return SUCCESS;
-}
-
-
 static int32_t delete_char_pkg(uint8_t *pkg, Command command, int32_t pkg_len) {
     // Required packet:
     // basic_header                [7]
@@ -649,6 +608,46 @@ static int32_t read_notepad_pkg(uint8_t *pkg, Command command, int32_t pkg_len) 
 
     pkg[11] = chk_bytes[0];
     pkg[12] = chk_bytes[1];
+
+    free(chk_bytes);
+
+    return SUCCESS;
+}
+
+static int32_t aura_led_config(uint8_t *pkg, Command command, int32_t pkg_len) {
+    // Required packet:
+    // basic_header                         [7]
+    // length    | 0x00 0x07                [2]
+    // instr     | 0x35                     [1]
+    // control   | aura_led_config.ctrl     [1]
+    // speed     | aura_led_config.speed    [1]
+    // color     | aura_led_config.color    [1]
+    // times     | aura_led_config.times    [1]
+    // chksum    | checksum                 [2]
+    /* Package length */
+    pkg[7] = 0x00;
+    pkg[8] = 0x07;
+
+    /* Instruction code */
+    pkg[9] = 0x35;
+
+    /* Control code */
+    pkg[10] = command.body.aura_led_config.ctrl;
+
+    /* Speed */
+    pkg[11] = command.body.aura_led_config.speed;
+
+    /* Color */
+    pkg[12] = command.body.aura_led_config.color;
+
+    /* Times */
+    pkg[13] = command.body.aura_led_config.times;
+
+    /* Checksum */
+    uint16_t chk = checksum(pkg, 6, pkg_len - 2);
+    uint8_t *chk_bytes = to_bytes_MSB(&chk, CHECKSUM_LEN);
+    pkg[14] = chk_bytes[0];
+    pkg[15] = chk_bytes[1];
 
     free(chk_bytes);
 

--- a/src/r502_driver.c
+++ b/src/r502_driver.c
@@ -112,7 +112,6 @@ int32_t call_cmd(Driver *driver, CommandType type, Reply *reply, int32_t arg_num
     if (err != SUCCESS || reply->conf_code != SUCCESS) {
         if (reply->conf_code != SUCCESS)
             err = reply->conf_code;
-
         goto error;
     }
 

--- a/src/r502_driver_utils.c
+++ b/src/r502_driver_utils.c
@@ -141,9 +141,6 @@ uint32_t get_command_reply_len(CommandType type) {
     case Store:
         len = STORE_REPLY_LEN;
         break;
-    case AuraLedConfig:
-        len = AURA_LED_CONFIG_REPLY_LEN;
-        break;
     case DeleteChar:
         len = DELETE_CHAR_REPLY_LEN;
         break;
@@ -155,6 +152,9 @@ uint32_t get_command_reply_len(CommandType type) {
         break;
     case ReadNotepad:
         len = READ_NOTEPAD_REPLY_LEN;
+        break;
+    case AuraLedConfig:
+        len = AURA_LED_CONFIG_REPLY_LEN;
         break;
     }
 

--- a/src/r502_driver_utils.c
+++ b/src/r502_driver_utils.c
@@ -99,6 +99,9 @@ uint32_t get_command_pkg_len(CommandType type) {
     case ReadNotepad:
         len = READ_NOTEPAD_LEN;
         break;
+    case AuraLedConfig:
+        len = AURA_LED_CONFIG_LEN;
+        break;
     }
 
     return len;
@@ -137,6 +140,9 @@ uint32_t get_command_reply_len(CommandType type) {
         break;
     case Store:
         len = STORE_REPLY_LEN;
+        break;
+    case AuraLedConfig:
+        len = AURA_LED_CONFIG_REPLY_LEN;
         break;
     case DeleteChar:
         len = DELETE_CHAR_REPLY_LEN;

